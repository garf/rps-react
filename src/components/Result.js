import React, {Component} from 'react';
import {Transition} from 'react-transition-group';
import './Result.scss';
import _ from 'lodash';
import ChoiceIcon from '@c/ChoiceIcon'

const animationDuration = 300;
const showDelay = 4000;
let timeout = null;

const defaultStyle = {
  transition: `opacity ${animationDuration}ms ease-in-out`,
  opacity: 1,
}

const transitionStyles = {
  exiting:  { opacity: 1 },
  exited:  { opacity: 0 },
};

const WinningPlayers = ({result}) =>
  _.some(result, entry => entry.result !== 0)
    ? result.map(entry => entry.result !== 0 ? <div key={entry.name}>{entry.name} Wins</div> : false)
    : 'DRAW';

class Result extends Component {
  constructor(props) {
    super(props);
    this.state = {
      display: false,
      result: [],
    };
  }

  showResult(result) {
    this.setState({display: true, result});
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      this.setState({display: false});
    }, showDelay);
  }

  render() {
    const {result} = this.state;

    return (
      <div className="result">
        <h3 className="has-text-centered is-size-3">Results</h3>
        <Transition in={this.state.display} timeout={animationDuration}>
          {state => (
            <div className="result__wrapper" style={{...defaultStyle, ...transitionStyles[state]}}>

              <div className="result__selections selections">
                {result.map(entry => (
                  <div key={entry.name} className="selections__choice">
                    <div className="selections__name">{entry.name}</div>
                    <div className={`selections__icon ${entry.result !== 0 ? 'has-text-success' : 'has-text-danger'}`}>
                      <ChoiceIcon choice={entry.choice} size={2} color={entry.result !== 0 ? 'green' : 'red'} />
                    </div>
                  </div>
                ))}
              </div>

              <hr />

              <div className="result__players">
                <WinningPlayers result={result} />
              </div>

            </div>
          )}
        </Transition>
      </div>
    );
  }
}

export {WinningPlayers, Result};

export default Result;
