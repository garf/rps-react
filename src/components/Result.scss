.result {
  &__selections {
    display: flex;
    justify-content: space-evenly;
  }

  &__players {
    display: flex;
    justify-content: center;
    color: #000;
  }
}

.selections {
  &__name {
    text-align: center;
  }

  &__icon {
    text-align: center;
  }
}
