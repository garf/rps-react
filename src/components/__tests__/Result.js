import React from 'react';
import {shallow} from 'enzyme';
import {Result, WinningPlayers} from '@c/Result';
import {CHOICE} from '@/constants'

describe('Result component', () => {

  it('should render correct ', () => {
    const component = shallow(<Result />);

    expect(component).toMatchSnapshot();
  });

  describe('WinningPlayers sub-component', () => {


    it('should render correct win message for player 1', () => {
      const result = [
        {
          name: 'User 1',
          result: 1,
          choice: CHOICE.PAPER,
        },
        {
          name: 'User 2',
          result: 0,
          choice: CHOICE.ROCK,
        },
      ];

      const component = shallow(<WinningPlayers result={result} />);

      expect(component).toMatchSnapshot();
    });

    it('should render correct win message for player 2', () => {
      const result = [
        {
          name: 'User 1',
          result: 0,
          choice: CHOICE.SCISSORS,
        },
        {
          name: 'User 2',
          result: 1,
          choice: CHOICE.ROCK,
        },
      ];

      const component = shallow(<WinningPlayers result={result} />);

      expect(component).toMatchSnapshot();
    });

    it('should render correct draw message', () => {
      const result = [
        {
          result: 0,
        },
        {
          result: 0,
        },
      ];

      const component = shallow(<WinningPlayers result={result} />);

      expect(component).toMatchSnapshot();
    });
  });
});
