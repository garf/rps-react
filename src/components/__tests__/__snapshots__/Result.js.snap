// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result component WinningPlayers sub-component should render correct draw message 1`] = `"DRAW"`;

exports[`Result component WinningPlayers sub-component should render correct win message for player 1 1`] = `
Array [
  <div
    key="User 1"
  >
    User 1
     Wins
  </div>,
  "",
]
`;

exports[`Result component WinningPlayers sub-component should render correct win message for player 2 1`] = `
Array [
  "",
  <div
    key="User 2"
  >
    User 2
     Wins
  </div>,
]
`;

exports[`Result component should render correct  1`] = `
<div
  className="result"
>
  <h3
    className="has-text-centered is-size-3"
  >
    Results
  </h3>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={false}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  >
    <Component />
  </Transition>
</div>
`;
